rasterio/fill.py

```diff
@@ -49,6 +50,16 @@ def fillnodata(
     out : numpy ndarray
         The filled raster array.
     """
+    if mask is None and hasattr(image, 'mask'):  # pragma: no cover
+        mask = ~image.mask
+    if not dtypes.is_ndarray(mask):
+        raise ValueError("An mask array is required")
+
+    if hasattr(image, 'mask'):  # pragma: no cover
+        image = image.data
+    if not dtypes.is_ndarray(image):
```

Comment:
This should move to `_fill.pyx` so that basic input validation is done at the lowest level.